### PR TITLE
add FasterRcnn(R50C4)

### PR DESCRIPTION
**Pull request by freedenS**
FasterRcnn is working.
MaskRcnn is coming soon.

### CODE REVIEW

File: rcnn/BatchedNms.cu [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+	const int *indices, float *scores, const float *classes, const float4 *boxes) {
+
+	// Go through detections by descending score
+	for (int m = 0; m < num_detections; m++) {
+		for (int n = 0; n < num_per_thread; n++) {
+			int i = threadIdx.x * num_per_thread + n;
+			if (i < num_detections && m < i && scores[m] > 0.0f) {
+				int idx = indices[i];
+				int max_idx = indices[m];
+				int icls = classes[idx];
```
Comment (wang-xinyu) [inline rcnn/BatchedNms.cu:31]: can you 4-space instead of tab? for all.

File: rcnn/README.md [hunk 2]
```diff
@@ -0,0 +1,71 @@
+# Rcnn
+
+The Pytorch implementation is [facebookresearch/detectron2](https://github.com/facebookresearch/detectron2).
+
+## Models
+
+- [x] Faster R-CNN(R50-C4)
+
+- [ ] Mask R-CNN(R50-C4)
+
```
Comment (wang-xinyu) [inline rcnn/README.md:10]: do you mean Faster RCNN not works for now?

**Comment by wang-xinyu** [comment]
Cool, thanks, just few comments.

File: rcnn/BatchedNms.cu (hunk 1 above)
Comment (freedenS) [reply rcnn/BatchedNms.cu:31]: Thanks, i'll adjust that.

File: rcnn/README.md (hunk 2 above)
Comment (freedenS) [reply rcnn/README.md:10]: sorry, i made a mistake.
I mean, faster-rcnn is working

**Review by wang-xinyu — CHANGES_REQUESTED** [review]
just some minor comments for coding style.

And how many plugins did you implement, can you make a simple explanation in readme?

File: rcnn/RoiAlign.cu [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+                const float x = roi_start_w + pw * bin_size_w +
+                    static_cast<float>(ix + .5f) * bin_size_w /
+                    static_cast<float>(roi_bin_grid_w);
+
+                float val = bilinear_interpolate(
+                    offset_bottom_data, height, width, y, x);
+
+                output_val += val;
+            }
+        }
```
Comment (wang-xinyu) [inline rcnn/RoiAlign.cu:139]: can you use the format like below for all?

```
if () {
} else {
}

for () {
}
```

**Comment by freedenS** [comment]
Thanks for your advice! I'll check and update soon. 😊

File: rcnn/BatchedNms.cu [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+#include <vector>
+#include <cmath>
+
+#include <cuda.h>
+#include <thrust/device_ptr.h>
+#include <thrust/sequence.h>
+#include <thrust/execution_policy.h>
+#include <thrust/gather.h>
+#include <thrust/system/cuda/detail/cub/device/device_radix_sort.cuh>
+
```
Comment (wang-xinyu) [inline rcnn/BatchedNms.cu:17]: what is `thrust`? do we need extra steps to install it?

Comment (freedenS) [reply rcnn/BatchedNms.cu:17]: No, just need to install cuda. thrust is located in the cuda include directory

Comment (wang-xinyu) [reply rcnn/BatchedNms.cu:17]: OK, cool.

**Comment by wang-xinyu** [comment]
BTW, do you have plan to implement maskRCNN and when can we have it ?

**Review by wang-xinyu — APPROVED** [review]

**Comment by freedenS** [comment]
> BTW, do you have plan to implement maskRCNN and when can we have it ?

Yeah, it will be completed next week.

**Pull request merged**